src/buildercore/threadbare/operations.py

```diff
@@ -0,0 +1,638 @@
+
+        temp_file, bytes_buffer = None, None
+        if hasattr(local_path, "read"):
+            # given a file-like object to download file into.
+            # 1. write the remote file to local temporary file
+            # 2. read temporary file into the given buffer
+            # 3. delete the temporary file
+
+            bytes_buffer = local_path
+            temp_file, local_path = tempfile.mkstemp(suffix="-threadbare")
+
+        if not os.path.isabs(local_path):
+            local_path = os.path.abspath(local_path)
+
+        if os.path.isdir(local_path):
+            local_path = os.path.join(local_path, os.path.basename(remote_path))
+
+        if use_sudo:
+            local_path = _download_as_root_hack(remote_path, local_path, **kwargs)
+
+        else:
+            if not remote_file_exists(remote_path, **kwargs):
+                raise EnvironmentError(
+                    "remote file does not exist: %s" % (remote_path,)
+                )
+            client = _ssh_client(**kwargs)
+            client.copy_remote_file(remote_path, local_path)
+
+        if temp_file:
+            bytes_buffer.write(open(local_path, "rb").read())
+            os.unlink(local_path)
+            return bytes_buffer
+
+        return local_path
+
+
+def _upload_as_root_hack(local_path, remote_path, **kwargs):
+    """uploads file at `local_path` to a remote temporary file then moves the file to `remote_path` as root.
+    does not alter any permissions or attributes on the file"""
+
+    client = _ssh_client(**kwargs)
+
+    cmd = single_command(
+        [
+            # create a temporary file with the suffix '-threadbare'
+            'tempfile=$(mktemp --suffix "-threadbare")',
+            'echo "$tempfile"',
+        ]
+    )
+    result = remote(cmd, **kwargs)
+    remote_temp_path = result["stdout"][-1]
+    assert remote_file_exists(remote_temp_path, **kwargs)  # sanity check
+
+    client.copy_file(local_path, remote_temp_path)
+    move_file_into_place = 'mv "%s" "%s"' % (remote_temp_path, remote_path)
+    remote_sudo(move_file_into_place, **kwargs)
+    assert remote_file_exists(remote_path, use_sudo=True, **kwargs)
+
+
+def _write_bytes_to_temporary_file(local_path):
+    """if `local_path` is a file-like object, write the contents to an *actual* file and
+    return a pair of new local filename and a function that removes the temporary file when called."""
+    if hasattr(local_path, "read"):
+        # `local_path` is a file-like object
+        local_bytes = local_path
+        local_bytes.seek(0)  # reset internal pointer
+        temp_file, local_path = tempfile.mkstemp(suffix="-threadbare")
+        with os.fdopen(temp_file, "wb") as fh:
+            fh.write(local_bytes.getvalue())
+        cleanup = lambda: os.unlink(local_path)
+        return local_path, cleanup
+    return local_path, None
+
+
+def upload(local_path, remote_path, use_sudo=False, **kwargs):
```

Comment:
Hoping this file won't need much maintenance as it is a lot of (forked from Fabric/written/rewritten?) code

Reply:
> as it is a lot of (forked from Fabric/written/rewritten?) code

I wrote all of this myself. It wasn't a clean room reverse engineering process and I had to do a lot of code spelunking to figure out Fabric idioms we were relying on, but what we have here in Threadbare and these three modules is quite lean and straightforward. It has excellent test coverage, including of remote commands.

But yes - definitely a lot of code and ideas to digest and review at once. I don't think you should approve it until you can do so without hesitation or regret. I'm happy to take any part of it and explain it in-depth.